// vso-node-api/handlers/basiccreds.js
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE file in the project root for full license information.
"use strict";
var BasicCredentialHandler = (function () {
    function BasicCredentialHandler(username, password) {
        this.username = username;
        this.password = password;
    }
    // currently implements pre-authorization
    // TODO: support preAuth = false where it hooks on 401
    BasicCredentialHandler.prototype.prepareRequest = function (options) {
        options.headers['Authorization'] = 'Basic ' + new Buffer(this.username + ':' + this.password).toString('base64');
        options.headers['X-TFS-FedAuthRedirect'] = 'Suppress';
    };
    // This handler cannot handle 401
    BasicCredentialHandler.prototype.canHandleAuthentication = function (res) {
        return false;
    };
    BasicCredentialHandler.prototype.handleAuthentication = function (httpClient, protocol, options, objs, finalCallback) {
    };
    return BasicCredentialHandler;
}());
exports.BasicCredentialHandler = BasicCredentialHandler;
