// vso-node-api/interfaces/FeatureManagementInterfaces.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
(function (ContributedFeatureEnabledValue) {
    /**
     * The state of the feature is not set for the specified scope
     */
    ContributedFeatureEnabledValue[ContributedFeatureEnabledValue["Undefined"] = -1] = "Undefined";
    /**
     * The feature is disabled at the specified scope
     */
    ContributedFeatureEnabledValue[ContributedFeatureEnabledValue["Disabled"] = 0] = "Disabled";
    /**
     * The feature is enabled at the specified scope
     */
    ContributedFeatureEnabledValue[ContributedFeatureEnabledValue["Enabled"] = 1] = "Enabled";
})(exports.ContributedFeatureEnabledValue || (exports.ContributedFeatureEnabledValue = {}));
var ContributedFeatureEnabledValue = exports.ContributedFeatureEnabledValue;
exports.TypeInfo = {
    ContributedFeature: {
        fields: null
    },
    ContributedFeatureDefaultValueRule: {
        fields: null
    },
    ContributedFeatureEnabledValue: {
        enumValues: {
            "undefined": -1,
            "disabled": 0,
            "enabled": 1,
        }
    },
    ContributedFeatureSettingScope: {
        fields: null
    },
    ContributedFeatureState: {
        fields: null
    },
};
exports.TypeInfo.ContributedFeature.fields = {
    defaultValueRules: {
        isArray: true,
        typeInfo: exports.TypeInfo.ContributedFeatureDefaultValueRule
    },
    scopes: {
        isArray: true,
        typeInfo: exports.TypeInfo.ContributedFeatureSettingScope
    },
};
exports.TypeInfo.ContributedFeatureDefaultValueRule.fields = {};
exports.TypeInfo.ContributedFeatureSettingScope.fields = {};
exports.TypeInfo.ContributedFeatureState.fields = {
    scope: {
        typeInfo: exports.TypeInfo.ContributedFeatureSettingScope
    },
    state: {
        enumType: exports.TypeInfo.ContributedFeatureEnabledValue
    },
};
