// azure-arm-rest-v2/azure-arm-app-service-kudu.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const tl = require("azure-pipelines-task-lib/task");
const fs = require("fs");
const util = require("util");
const webClient = require("./webClient");
const constants_1 = require("./constants");
class KuduServiceManagementClient {
    constructor(scmUri, accessToken) {
        this._cookie = undefined;
        this._accesssToken = accessToken;
        this._scmUri = scmUri;
    }
    beginRequest(request, reqOptions) {
        return __awaiter(this, void 0, void 0, function* () {
            request.headers = request.headers || {};
            request.headers["Authorization"] = "Basic " + this._accesssToken;
            if (!request.headers['Content-Type']) {
                request.headers['Content-Type'] = 'application/json; charset=utf-8';
            }
            if (!!this._cookie) {
                tl.debug(`setting affinity cookie ${JSON.stringify(this._cookie)}`);
                request.headers['Cookie'] = this._cookie;
            }
            let retryCount = reqOptions && util.isNumber(reqOptions.retryCount) ? reqOptions.retryCount : 5;
            while (retryCount >= 0) {
                try {
                    let httpResponse = yield webClient.sendRequest(request, reqOptions);
                    if (httpResponse.headers['set-cookie'] && !this._cookie) {
                        this._cookie = httpResponse.headers['set-cookie'];
                        tl.debug(`loaded affinity cookie ${JSON.stringify(this._cookie)}`);
                    }
                    return httpResponse;
                }
                catch (exception) {
                    let exceptionString = exception.toString();
                    if (exceptionString.indexOf("Hostname/IP doesn't match certificates's altnames") != -1
                        || exceptionString.indexOf("unable to verify the first certificate") != -1
                        || exceptionString.indexOf("unable to get local issuer certificate") != -1) {
                        tl.warning(tl.loc('ASE_SSLIssueRecommendation'));
                    }
                    if (retryCount > 0 && exceptionString.indexOf('Request timeout') != -1 && (!reqOptions || reqOptions.retryRequestTimedout)) {
                        tl.debug('encountered request timedout issue in Kudu. Retrying again');
                        retryCount -= 1;
                        continue;
                    }
                    throw new Error(exceptionString);
                }
            }
        });
    }
    getRequestUri(uriFormat, queryParameters) {
        uriFormat = uriFormat[0] == "/" ? uriFormat : "/" + uriFormat;
        if (queryParameters && queryParameters.length > 0) {
            uriFormat = uriFormat + '?' + queryParameters.join('&');
        }
        return this._scmUri + uriFormat;
    }
    getScmUri() {
        return this._scmUri;
    }
}
exports.KuduServiceManagementClient = KuduServiceManagementClient;
class Kudu {
    constructor(scmUri, username, password) {
        var base64EncodedCredential = (new Buffer(username + ':' + password).toString('base64'));
        this._client = new KuduServiceManagementClient(scmUri, base64EncodedCredential);
    }
    updateDeployment(requestBody) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'PUT';
            httpRequest.body = JSON.stringify(requestBody);
            httpRequest.uri = this._client.getRequestUri(`/api/deployments/${requestBody.id}`);
            try {
                let webRequestOptions = { retriableErrorCodes: [], retriableStatusCodes: [], retryCount: 1, retryIntervalInSeconds: 5, retryRequestTimedout: true };
                var response = yield this._client.beginRequest(httpRequest, webRequestOptions);
                tl.debug(`updateDeployment. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    console.log(tl.loc("Successfullyupdateddeploymenthistory", response.body.url));
                    return response.body.id;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('Failedtoupdatedeploymenthistory', this._getFormattedError(error)));
            }
        });
    }
    getContinuousJobs() {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/continuouswebjobs`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getContinuousJobs. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetContinuousWebJobs', this._getFormattedError(error)));
            }
        });
    }
    startContinuousWebJob(jobName) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log(tl.loc('StartingWebJob', jobName));
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.uri = this._client.getRequestUri(`/api/continuouswebjobs/${jobName}/start`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`startContinuousWebJob. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    console.log(tl.loc('StartedWebJob', jobName));
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToStartContinuousWebJob', jobName, this._getFormattedError(error)));
            }
        });
    }
    stopContinuousWebJob(jobName) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log(tl.loc('StoppingWebJob', jobName));
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.uri = this._client.getRequestUri(`/api/continuouswebjobs/${jobName}/stop`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`stopContinuousWebJob. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    console.log(tl.loc('StoppedWebJob', jobName));
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToStopContinuousWebJob', jobName, this._getFormattedError(error)));
            }
        });
    }
    installSiteExtension(extensionID) {
        return __awaiter(this, void 0, void 0, function* () {
            console.log(tl.loc("InstallingSiteExtension", extensionID));
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'PUT';
            httpRequest.uri = this._client.getRequestUri(`/api/siteextensions/${extensionID}`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`installSiteExtension. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    console.log(tl.loc("SiteExtensionInstalled", extensionID));
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToInstallSiteExtension', extensionID, this._getFormattedError(error)));
            }
        });
    }
    getSiteExtensions() {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/siteextensions`, ['checkLatest=false']);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getSiteExtensions. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetSiteExtensions', this._getFormattedError(error)));
            }
        });
    }
    getAllSiteExtensions() {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/extensionfeed`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getAllSiteExtensions. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetAllSiteExtensions', this._getFormattedError(error)));
            }
        });
    }
    getProcess(processID) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/processes/${processID}`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getProcess. status code: ${response.statusCode} - ${response.statusMessage}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetProcess', this._getFormattedError(error)));
            }
        });
    }
    killProcess(processID) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'DELETE';
            httpRequest.uri = this._client.getRequestUri(`/api/processes/${processID}`);
            var reqOptions = {
                retriableErrorCodes: ["ETIMEDOUT"],
                retriableStatusCodes: [503],
                retryCount: 1,
                retryIntervalInSeconds: 5,
                retryRequestTimedout: true
            };
            try {
                var response = yield this._client.beginRequest(httpRequest, reqOptions);
                tl.debug(`killProcess. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 502) {
                    tl.debug(`Killed Process ${processID}`);
                    return;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToKillProcess', this._getFormattedError(error)));
            }
        });
    }
    getAppSettings() {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/settings`);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getAppSettings. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToFetchKuduAppSettings', this._getFormattedError(error)));
            }
        });
    }
    listDir(physicalPath) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}/`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`listFiles. Data: ${JSON.stringify(response)}`);
                if ([200, 201, 204].indexOf(response.statusCode) != -1) {
                    return response.body;
                }
                else if (response.statusCode === 404) {
                    return null;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(tl.loc('FailedToListPath', physicalPath, this._getFormattedError(error)));
            }
        });
    }
    getFileContent(physicalPath, fileName) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}/${fileName}`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getFileContent. Status code: ${response.statusCode} - ${response.statusMessage}`);
                if ([200, 201, 204].indexOf(response.statusCode) != -1) {
                    return response.body;
                }
                else if (response.statusCode === 404) {
                    return null;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetFileContent', physicalPath, fileName, this._getFormattedError(error)));
            }
        });
    }
    uploadFile(physicalPath, fileName, filePath) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            if (!tl.exist(filePath)) {
                throw new Error(tl.loc('FilePathInvalid', filePath));
            }
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'PUT';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}/${fileName}`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            httpRequest.body = fs.createReadStream(filePath);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`uploadFile. Data: ${JSON.stringify(response)}`);
                if ([200, 201, 204].indexOf(response.statusCode) != -1) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToUploadFile', physicalPath, fileName, this._getFormattedError(error)));
            }
        });
    }
    createPath(physicalPath) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'PUT';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}/`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`createPath. Data: ${JSON.stringify(response)}`);
                if ([200, 201, 204].indexOf(response.statusCode) != -1) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToCreatePath', physicalPath, this._getFormattedError(error)));
            }
        });
    }
    runCommand(physicalPath, command) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.uri = this._client.getRequestUri(`/api/command`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            httpRequest.body = JSON.stringify({
                'command': command,
                'dir': physicalPath
            });
            try {
                tl.debug('Executing Script on Kudu. Command: ' + command);
                let webRequestOptions = { retriableErrorCodes: null, retriableStatusCodes: null, retryCount: 5, retryIntervalInSeconds: 5, retryRequestTimedout: false };
                var response = yield this._client.beginRequest(httpRequest, webRequestOptions);
                tl.debug(`runCommand. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(error.toString());
            }
        });
    }
    extractZIP(webPackage, physicalPath) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'PUT';
            httpRequest.uri = this._client.getRequestUri(`/api/zip/${physicalPath}/`);
            httpRequest.headers = {
                'Content-Type': 'multipart/form-data',
                'If-Match': '*'
            };
            httpRequest.body = fs.createReadStream(webPackage);
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`extractZIP. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(tl.loc('Failedtodeploywebapppackageusingkuduservice', this._getFormattedError(error)));
            }
        });
    }
    zipDeploy(webPackage, queryParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            let httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.uri = this._client.getRequestUri(`/api/zipdeploy`, queryParameters);
            httpRequest.body = fs.createReadStream(webPackage);
            httpRequest.headers = {
                'Content-Type': 'application/octet-stream'
            };
            try {
                let response = yield this._client.beginRequest(httpRequest);
                tl.debug(`ZIP Deploy response: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    tl.debug('Deployment passed');
                    return null;
                }
                else if (response.statusCode == 202) {
                    let pollableURL = response.headers.location;
                    if (!!pollableURL) {
                        tl.debug(`Polling for ZIP Deploy URL: ${pollableURL}`);
                        return yield this._getDeploymentDetailsFromPollURL(pollableURL);
                    }
                    else {
                        tl.debug('zip deploy returned 202 without pollable URL.');
                        return null;
                    }
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw new Error(tl.loc('PackageDeploymentFailed', this._getFormattedError(error)));
            }
        });
    }
    warDeploy(webPackage, queryParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            let httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.uri = this._client.getRequestUri(`/api/wardeploy`, queryParameters);
            httpRequest.body = fs.createReadStream(webPackage);
            httpRequest.headers = {
                'Content-Type': 'application/octet-stream'
            };
            try {
                let response = yield this._client.beginRequest(httpRequest);
                tl.debug(`War Deploy response: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    tl.debug('Deployment passed');
                    return null;
                }
                else if (response.statusCode == 202) {
                    let pollableURL = response.headers.location;
                    if (!!pollableURL) {
                        tl.debug(`Polling for War Deploy URL: ${pollableURL}`);
                        return yield this._getDeploymentDetailsFromPollURL(pollableURL);
                    }
                    else {
                        tl.debug('war deploy returned 202 without pollable URL.');
                        return null;
                    }
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw new Error(tl.loc('PackageDeploymentFailed', this._getFormattedError(error)));
            }
        });
    }
    getDeploymentDetails(deploymentID) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                var httpRequest = new webClient.WebRequest();
                httpRequest.method = 'GET';
                httpRequest.uri = this._client.getRequestUri(`/api/deployments/${deploymentID}`);
                ;
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getDeploymentDetails. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetDeploymentLogs', this._getFormattedError(error)));
            }
        });
    }
    getDeploymentLogs(log_url) {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                var httpRequest = new webClient.WebRequest();
                httpRequest.method = 'GET';
                httpRequest.uri = log_url;
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`getDeploymentLogs. Data: ${JSON.stringify(response)}`);
                if (response.statusCode == 200) {
                    return response.body;
                }
                throw response;
            }
            catch (error) {
                throw Error(tl.loc('FailedToGetDeploymentLogs', this._getFormattedError(error)));
            }
        });
    }
    deleteFile(physicalPath, fileName) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'DELETE';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}/${fileName}`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`deleteFile. Data: ${JSON.stringify(response)}`);
                if ([200, 201, 204, 404].indexOf(response.statusCode) != -1) {
                    return;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(tl.loc('FailedToDeleteFile', physicalPath, fileName, this._getFormattedError(error)));
            }
        });
    }
    deleteFolder(physicalPath) {
        return __awaiter(this, void 0, void 0, function* () {
            physicalPath = physicalPath.replace(/[\\]/g, "/");
            physicalPath = physicalPath[0] == "/" ? physicalPath.slice(1) : physicalPath;
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'DELETE';
            httpRequest.uri = this._client.getRequestUri(`/api/vfs/${physicalPath}`);
            httpRequest.headers = {
                'If-Match': '*'
            };
            try {
                var response = yield this._client.beginRequest(httpRequest);
                tl.debug(`deleteFolder. Data: ${JSON.stringify(response)}`);
                if ([200, 201, 204, 404].indexOf(response.statusCode) != -1) {
                    return;
                }
                else {
                    throw response;
                }
            }
            catch (error) {
                throw Error(tl.loc('FailedToDeleteFolder', physicalPath, this._getFormattedError(error)));
            }
        });
    }
    _getDeploymentDetailsFromPollURL(pollURL) {
        return __awaiter(this, void 0, void 0, function* () {
            let httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.uri = pollURL;
            httpRequest.headers = {};
            while (true) {
                let response = yield this._client.beginRequest(httpRequest);
                if (response.statusCode == 200 || response.statusCode == 202) {
                    var result = response.body;
                    tl.debug(`POLL URL RESULT: ${JSON.stringify(response)}`);
                    if (result.status == constants_1.KUDU_DEPLOYMENT_CONSTANTS.SUCCESS || result.status == constants_1.KUDU_DEPLOYMENT_CONSTANTS.FAILED) {
                        return result;
                    }
                    else {
                        tl.debug(`Deployment status: ${result.status} '${result.status_text}'. retry after 5 seconds`);
                        yield webClient.sleepFor(5);
                        continue;
                    }
                }
                else {
                    throw response;
                }
            }
        });
    }
    _getFormattedError(error) {
        if (error && error.statusCode) {
            return `${error.statusMessage} (CODE: ${error.statusCode})`;
        }
        else if (error && error.message) {
            if (error.statusCode) {
                error.message = `${typeof error.message.valueOf() == 'string' ? error.message : error.message.Code + " - " + error.message.Message} (CODE: ${error.statusCode})`;
            }
            return error.message;
        }
        return error;
    }
}
exports.Kudu = Kudu;
