// azure-arm-rest-v2/azure-arm-storage.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const azureServiceClient = require("./AzureServiceClient");
const webClient = require("./webClient");
const tl = require("azure-pipelines-task-lib/task");
const Q = require("q");
class StorageManagementClient extends azureServiceClient.ServiceClient {
    constructor(credentials, subscriptionId, baseUri, options) {
        super(credentials, subscriptionId);
        this.acceptLanguage = 'en-US';
        this.generateClientRequestId = true;
        this.apiVersion = (credentials.isAzureStackEnvironment) ? '2015-06-15' : '2017-06-01';
        if (!options)
            options = {};
        if (baseUri) {
            this.baseUri = baseUri;
        }
        if (options.acceptLanguage) {
            this.acceptLanguage = options.acceptLanguage;
        }
        if (options.longRunningOperationRetryTimeout) {
            this.longRunningOperationRetryTimeout = options.longRunningOperationRetryTimeout;
        }
        if (options.generateClientRequestId) {
            this.generateClientRequestId = options.generateClientRequestId;
        }
        this.storageAccounts = new StorageAccounts(this);
    }
}
exports.StorageManagementClient = StorageManagementClient;
class StorageAccounts {
    constructor(client) {
        this.client = client;
    }
    list(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.headers = this.client.setCustomHeaders(options);
            // Getting all azure rm storage accounts (along with resource group names) for the given subscription.
            httpRequest.uri = this.client.getRequestUri('//subscriptions/{subscriptionId}/providers/Microsoft.Storage/storageAccounts', {});
            var deferred = Q.defer();
            var result = [];
            this.client.beginRequest(httpRequest).then((response) => __awaiter(this, void 0, void 0, function* () {
                if (response.statusCode == 200) {
                    if (response.body.value) {
                        let storageAccounts = response.body.value;
                        result = result.concat(storageAccounts);
                    }
                    if (response.body.nextLink) {
                        var nextResult = yield this.client.accumulateResultFromPagedResult(response.body.nextLink);
                        if (nextResult.error) {
                            deferred.reject(nextResult.error);
                        }
                        let storageAccounts = nextResult.result;
                        result = result.concat(storageAccounts);
                    }
                    deferred.resolve(result);
                }
                else {
                    deferred.reject(azureServiceClient.ToError(response));
                }
            })).catch(function (error) {
                deferred.reject(error);
            });
            return deferred.promise;
        });
    }
    listClassicAndRMAccounts(options) {
        return __awaiter(this, void 0, void 0, function* () {
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'GET';
            httpRequest.headers = this.client.setCustomHeaders(options);
            // Getting all storage accounts (azure rm and classic, along with resource group names) for the given subscription.
            httpRequest.uri = "https://management.azure.com/resources?api-version=2014-04-01-preview&%24filter=(subscriptionId%20eq%20'{subscriptionId}')%20and%20(resourceType%20eq%20'microsoft.storage%2Fstorageaccounts'%20or%20resourceType%20eq%20'microsoft.classicstorage%2Fstorageaccounts')";
            httpRequest.uri = httpRequest.uri.replace('{subscriptionId}', this.client.subscriptionId);
            var deferred = Q.defer();
            var result = [];
            this.client.beginRequest(httpRequest).then((response) => __awaiter(this, void 0, void 0, function* () {
                if (response.statusCode == 200) {
                    if (response.body.value) {
                        let storageAccounts = response.body.value;
                        result = result.concat(storageAccounts);
                    }
                    if (response.body.nextLink) {
                        var nextResult = yield this.client.accumulateResultFromPagedResult(response.body.nextLink);
                        if (nextResult.error) {
                            deferred.reject(nextResult.error);
                        }
                        let storageAccounts = nextResult.result;
                        result = result.concat(storageAccounts);
                    }
                    deferred.resolve(result);
                }
                else {
                    deferred.reject(azureServiceClient.ToError(response));
                }
            })).catch(function (error) {
                deferred.reject(error);
            });
            return deferred.promise;
        });
    }
    listKeys(resourceGroupName, accountName, options, storageAccountType) {
        return __awaiter(this, void 0, void 0, function* () {
            if (resourceGroupName === null || resourceGroupName === undefined || typeof resourceGroupName.valueOf() !== 'string') {
                throw new Error(tl.loc("ResourceGroupCannotBeNull"));
            }
            if (accountName === null || accountName === undefined || typeof accountName.valueOf() !== 'string') {
                throw new Error(tl.loc("StorageAccountCannotBeNull"));
            }
            var apiVersion = "2017-06-01";
            var resourceProvider = "Microsoft.Storage";
            if (!!storageAccountType && storageAccountType.toLowerCase().indexOf("classicstorage") > 0) {
                resourceProvider = "Microsoft.ClassicStorage";
                apiVersion = "2015-12-01";
            }
            var httpRequest = new webClient.WebRequest();
            httpRequest.method = 'POST';
            httpRequest.headers = this.client.setCustomHeaders(options);
            httpRequest.uri = this.client.getRequestUri('//subscriptions/{subscriptionId}/resourcegroups/{resourceGroupName}/providers/{provider}/storageAccounts/{storageAccountName}/listKeys', {
                '{resourceGroupName}': resourceGroupName,
                '{storageAccountName}': accountName,
                '{provider}': resourceProvider
            }, [], apiVersion);
            var deferred = Q.defer();
            var accessKeys = [];
            this.client.beginRequest(httpRequest).then((response) => {
                if (response.statusCode == 200) {
                    if (resourceProvider === "Microsoft.ClassicStorage") {
                        accessKeys[0] = response.body.primaryKey;
                        accessKeys[1] = response.body.secondaryKey;
                    }
                    else if (response.body.keys) {
                        let keys = response.body.keys;
                        for (let i = 0; i < keys.length; i++) {
                            accessKeys[i] = keys[i]["value"];
                        }
                    }
                    deferred.resolve(accessKeys);
                }
                else {
                    deferred.reject(azureServiceClient.ToError(response));
                }
            }).catch(function (error) {
                deferred.reject(error);
            });
            return deferred.promise;
        });
    }
    get(storageAccountName) {
        return __awaiter(this, void 0, void 0, function* () {
            let storageAccounts = yield this.list(null);
            let index = storageAccounts.findIndex(account => account.name.toLowerCase() === storageAccountName.toLowerCase());
            if (index < 0) {
                throw new Error(tl.loc("StorageAccountDoesNotExist", storageAccountName));
            }
            return storageAccounts[index];
        });
    }
    static getResourceGroupNameFromUri(resourceUri) {
        if (this.isNonEmptyInternal(resourceUri)) {
            resourceUri = resourceUri.toLowerCase();
            return resourceUri.substring(resourceUri.indexOf("resourcegroups/") + "resourcegroups/".length, resourceUri.indexOf("/providers"));
        }
        return "";
    }
    static isNonEmptyInternal(str) {
        return (!!str && !!str.trim());
    }
}
exports.StorageAccounts = StorageAccounts;
