// vso-node-api/BuildApi.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
// Licensed under the MIT license.  See LICENSE file in the project root for full license information.
var Q = require('q');
var basem = require('./ClientApiBases');
var BuildInterfaces = require("./interfaces/BuildInterfaces");
var BuildApi = (function (_super) {
    __extends(BuildApi, _super);
    function BuildApi(baseUrl, handlers) {
        _super.call(this, baseUrl, handlers, 'node-Build-api');
    }
    /**
    * Associates an artifact with a build
    *
    * @param {BuildInterfaces.BuildArtifact} artifact
    * @param {number} buildId
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.createArtifact = function (artifact, buildId, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, artifact) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(artifact);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "1db06c96-014e-44e1-ac91-90b2d4b3e984", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.create(url, apiVersion, artifact, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets a specific artifact for a build
    *
    * @param {number} buildId
    * @param {string} artifactName
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getArtifact = function (buildId, artifactName, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, artifact) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(artifact);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            artifactName: artifactName,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "1db06c96-014e-44e1-ac91-90b2d4b3e984", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets a specific artifact for a build
    *
    * @param {number} buildId
    * @param {string} artifactName
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getArtifactContentZip = function (buildId, artifactName, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, artifact) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(artifact);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            artifactName: artifactName,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "1db06c96-014e-44e1-ac91-90b2d4b3e984", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("application/zip", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets all artifacts for a build
    *
    * @param {number} buildId
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getArtifacts = function (buildId, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, artifacts) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(artifacts);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "1db06c96-014e-44e1-ac91-90b2d4b3e984", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project
    * @param {number} definitionId
    * @param {string} branchName
    */
    BuildApi.prototype.getBadge = function (project, definitionId, branchName) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, badge) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(badge);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        var queryValues = {
            branchName: branchName,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "de6a4df8-22cd-44ee-af2d-39f6aa7a4261", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {string} repoType
    * @param {string} repoId
    * @param {string} branchName
    */
    BuildApi.prototype.getBuildBadge = function (project, repoType, repoId, branchName) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, buildbadge) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(buildbadge);
            }
        };
        var routeValues = {
            project: project,
            repoType: repoType
        };
        var queryValues = {
            repoId: repoId,
            branchName: branchName,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "21b3b9ce-fad5-4567-9ad0-80679794e003", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {string} repoType
    * @param {string} repoId
    * @param {string} branchName
    */
    BuildApi.prototype.getBuildBadgeData = function (project, repoType, repoId, branchName) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, buildbadge) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(buildbadge);
            }
        };
        var routeValues = {
            project: project,
            repoType: repoType
        };
        var queryValues = {
            repoId: repoId,
            branchName: branchName,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "21b3b9ce-fad5-4567-9ad0-80679794e003", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes a build
    *
    * @param {number} buildId
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.deleteBuild = function (buildId, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Gets a build
    *
    * @param {number} buildId
    * @param {string} project - Project ID or project name
    * @param {string} propertyFilters - A comma-delimited list of properties to include in the results
    */
    BuildApi.prototype.getBuild = function (buildId, project, propertyFilters) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, build) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(build);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            propertyFilters: propertyFilters,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Build, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets builds
    *
    * @param {string} project - Project ID or project name
    * @param {number[]} definitions - A comma-delimited list of definition ids
    * @param {number[]} queues - A comma-delimited list of queue ids
    * @param {string} buildNumber
    * @param {Date} minFinishTime
    * @param {Date} maxFinishTime
    * @param {string} requestedFor
    * @param {BuildInterfaces.BuildReason} reasonFilter
    * @param {BuildInterfaces.BuildStatus} statusFilter
    * @param {BuildInterfaces.BuildResult} resultFilter
    * @param {string[]} tagFilters - A comma-delimited list of tags
    * @param {string[]} properties - A comma-delimited list of properties to include in the results
    * @param {number} top - The maximum number of builds to retrieve
    * @param {string} continuationToken
    * @param {number} maxBuildsPerDefinition
    * @param {BuildInterfaces.QueryDeletedOption} deletedFilter
    * @param {BuildInterfaces.BuildQueryOrder} queryOrder
    * @param {string} branchName
    * @param {number[]} buildIds
    * @param {string} repositoryId
    * @param {string} repositoryType
    */
    BuildApi.prototype.getBuilds = function (project, definitions, queues, buildNumber, minFinishTime, maxFinishTime, requestedFor, reasonFilter, statusFilter, resultFilter, tagFilters, properties, top, continuationToken, maxBuildsPerDefinition, deletedFilter, queryOrder, branchName, buildIds, repositoryId, repositoryType) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, builds) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(builds);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            definitions: definitions && definitions.join(","),
            queues: queues && queues.join(","),
            buildNumber: buildNumber,
            minFinishTime: minFinishTime,
            maxFinishTime: maxFinishTime,
            requestedFor: requestedFor,
            reasonFilter: reasonFilter,
            statusFilter: statusFilter,
            resultFilter: resultFilter,
            tagFilters: tagFilters && tagFilters.join(","),
            properties: properties && properties.join(","),
            '$top': top,
            continuationToken: continuationToken,
            maxBuildsPerDefinition: maxBuildsPerDefinition,
            deletedFilter: deletedFilter,
            queryOrder: queryOrder,
            branchName: branchName,
            buildIds: buildIds && buildIds.join(","),
            repositoryId: repositoryId,
            repositoryType: repositoryType,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Build, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Queues a build
    *
    * @param {BuildInterfaces.Build} build
    * @param {string} project - Project ID or project name
    * @param {boolean} ignoreWarnings
    * @param {string} checkInTicket
    */
    BuildApi.prototype.queueBuild = function (build, project, ignoreWarnings, checkInTicket) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, build) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(build);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            ignoreWarnings: ignoreWarnings,
            checkInTicket: checkInTicket,
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.Build, responseTypeMetadata: BuildInterfaces.TypeInfo.Build, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, build, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Updates a build
    *
    * @param {BuildInterfaces.Build} build
    * @param {number} buildId
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.updateBuild = function (build, buildId, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, build) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(build);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.Build, responseTypeMetadata: BuildInterfaces.TypeInfo.Build, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, build, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Update a batch of builds
    *
    * @param {BuildInterfaces.Build[]} builds
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.updateBuilds = function (builds, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, builds) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(builds);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.3", "build", "0cd358e1-9217-4d94-8269-1c1ee6f93dcf", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.Build, responseTypeMetadata: BuildInterfaces.TypeInfo.Build, responseIsCollection: true };
            _this.restClient.update(url, apiVersion, builds, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the changes associated with a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} continuationToken
    * @param {number} top - The maximum number of changes to return
    * @param {boolean} includeSourceChange
    */
    BuildApi.prototype.getBuildChanges = function (project, buildId, continuationToken, top, includeSourceChange) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, changes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(changes);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            continuationToken: continuationToken,
            '$top': top,
            includeSourceChange: includeSourceChange,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "54572c7b-bbd3-45d4-80dc-28be08941620", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Change, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the changes associated between given builds
    *
    * @param {string} project - Project ID or project name
    * @param {number} fromBuildId
    * @param {number} toBuildId
    * @param {number} top - The maximum number of changes to return
    */
    BuildApi.prototype.getChangesBetweenBuilds = function (project, fromBuildId, toBuildId, top) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, changes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(changes);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            fromBuildId: fromBuildId,
            toBuildId: toBuildId,
            '$top': top,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "f10f0ea5-18a1-43ec-a8fb-2042c7be9b43", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Change, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets a controller
    *
    * @param {number} controllerId
    */
    BuildApi.prototype.getBuildController = function (controllerId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, Controller) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(Controller);
            }
        };
        var routeValues = {
            controllerId: controllerId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "fcac1932-2ee1-437f-9b6f-7f696be858f6", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildController, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets controller, optionally filtered by name
    *
    * @param {string} name
    */
    BuildApi.prototype.getBuildControllers = function (name) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, Controllers) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(Controllers);
            }
        };
        var routeValues = {};
        var queryValues = {
            name: name,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "fcac1932-2ee1-437f-9b6f-7f696be858f6", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildController, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Creates a new definition
    *
    * @param {BuildInterfaces.BuildDefinition} definition
    * @param {string} project - Project ID or project name
    * @param {number} definitionToCloneId
    * @param {number} definitionToCloneRevision
    */
    BuildApi.prototype.createDefinition = function (definition, project, definitionToCloneId, definitionToCloneRevision) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, definition) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(definition);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            definitionToCloneId: definitionToCloneId,
            definitionToCloneRevision: definitionToCloneRevision,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "dbeaf647-6167-421a-bda9-c9327b25e2e6", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinition, responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinition, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, definition, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes a definition and all associated builds
    *
    * @param {number} definitionId
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.deleteDefinition = function (definitionId, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "dbeaf647-6167-421a-bda9-c9327b25e2e6", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Gets a definition, optionally at a specific revision
    *
    * @param {number} definitionId
    * @param {string} project - Project ID or project name
    * @param {number} revision
    * @param {string[]} propertyFilters
    */
    BuildApi.prototype.getDefinition = function (definitionId, project, revision, propertyFilters) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, definition) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(definition);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        var queryValues = {
            revision: revision,
            propertyFilters: propertyFilters && propertyFilters.join(","),
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "dbeaf647-6167-421a-bda9-c9327b25e2e6", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinition, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets definitions, optionally filtered by name
    *
    * @param {string} project - Project ID or project name
    * @param {string} name
    * @param {string} repositoryId
    * @param {string} repositoryType
    * @param {BuildInterfaces.DefinitionQueryOrder} queryOrder
    * @param {number} top
    * @param {string} continuationToken
    * @param {Date} minMetricsTime
    * @param {number[]} definitionIds
    * @param {string} path
    * @param {Date} builtAfter
    * @param {Date} notBuiltAfter
    */
    BuildApi.prototype.getDefinitions = function (project, name, repositoryId, repositoryType, queryOrder, top, continuationToken, minMetricsTime, definitionIds, path, builtAfter, notBuiltAfter) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, definitions) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(definitions);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            name: name,
            repositoryId: repositoryId,
            repositoryType: repositoryType,
            queryOrder: queryOrder,
            '$top': top,
            continuationToken: continuationToken,
            minMetricsTime: minMetricsTime,
            definitionIds: definitionIds && definitionIds.join(","),
            path: path,
            builtAfter: builtAfter,
            notBuiltAfter: notBuiltAfter,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "dbeaf647-6167-421a-bda9-c9327b25e2e6", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionReference, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Updates an existing definition
    *
    * @param {BuildInterfaces.BuildDefinition} definition
    * @param {number} definitionId
    * @param {string} project - Project ID or project name
    * @param {number} secretsSourceDefinitionId
    * @param {number} secretsSourceDefinitionRevision
    */
    BuildApi.prototype.updateDefinition = function (definition, definitionId, project, secretsSourceDefinitionId, secretsSourceDefinitionRevision) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, definition) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(definition);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        var queryValues = {
            secretsSourceDefinitionId: secretsSourceDefinitionId,
            secretsSourceDefinitionRevision: secretsSourceDefinitionRevision,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "dbeaf647-6167-421a-bda9-c9327b25e2e6", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinition, responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinition, responseIsCollection: false };
            _this.restClient.replace(url, apiVersion, definition, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Creates a new folder
    *
    * @param {BuildInterfaces.Folder} folder
    * @param {string} project - Project ID or project name
    * @param {string} path
    */
    BuildApi.prototype.createFolder = function (folder, project, path) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, folder) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(folder);
            }
        };
        var routeValues = {
            project: project,
            path: path
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "a906531b-d2da-4f55-bda7-f3e676cc50d9", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.Folder, responseTypeMetadata: BuildInterfaces.TypeInfo.Folder, responseIsCollection: false };
            _this.restClient.replace(url, apiVersion, folder, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes a definition folder for given folder name and path and all it's existing definitions and it's corresponding builds
    *
    * @param {string} project - Project ID or project name
    * @param {string} path
    */
    BuildApi.prototype.deleteFolder = function (project, path) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            path: path
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "a906531b-d2da-4f55-bda7-f3e676cc50d9", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Gets folders
    *
    * @param {string} project - Project ID or project name
    * @param {string} path
    * @param {BuildInterfaces.FolderQueryOrder} queryOrder
    */
    BuildApi.prototype.getFolders = function (project, path, queryOrder) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, folders) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(folders);
            }
        };
        var routeValues = {
            project: project,
            path: path
        };
        var queryValues = {
            queryOrder: queryOrder,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "a906531b-d2da-4f55-bda7-f3e676cc50d9", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Folder, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Updates an existing folder at given  existing path
    *
    * @param {BuildInterfaces.Folder} folder
    * @param {string} project - Project ID or project name
    * @param {string} path
    */
    BuildApi.prototype.updateFolder = function (folder, project, path) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, folder) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(folder);
            }
        };
        var routeValues = {
            project: project,
            path: path
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "a906531b-d2da-4f55-bda7-f3e676cc50d9", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.Folder, responseTypeMetadata: BuildInterfaces.TypeInfo.Folder, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, folder, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets a log
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {number} logId
    * @param {number} startLine
    * @param {number} endLine
    */
    BuildApi.prototype.getBuildLog = function (project, buildId, logId, startLine, endLine) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, log) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(log);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId,
            logId: logId
        };
        var queryValues = {
            startLine: startLine,
            endLine: endLine,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "35a80daf-7f30-45fc-86e8-6b813d9c90df", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("text/plain", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets a log
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {number} logId
    * @param {number} startLine
    * @param {number} endLine
    */
    BuildApi.prototype.getBuildLogLines = function (project, buildId, logId, startLine, endLine) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, logs) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(logs);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId,
            logId: logId
        };
        var queryValues = {
            startLine: startLine,
            endLine: endLine,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "35a80daf-7f30-45fc-86e8-6b813d9c90df", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets logs for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    */
    BuildApi.prototype.getBuildLogs = function (project, buildId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, logs) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(logs);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "35a80daf-7f30-45fc-86e8-6b813d9c90df", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildLog, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets logs for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    */
    BuildApi.prototype.getBuildLogsZip = function (project, buildId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, log) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(log);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "35a80daf-7f30-45fc-86e8-6b813d9c90df", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("application/zip", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets metrics of a definition
    *
    * @param {string} project - Project ID or project name
    * @param {number} definitionId
    * @param {Date} minMetricsTime
    */
    BuildApi.prototype.getDefinitionMetrics = function (project, definitionId, minMetricsTime) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, metrics) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(metrics);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        var queryValues = {
            minMetricsTime: minMetricsTime,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "d973b939-0ce0-4fec-91d8-da3940fa1827", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildMetric, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getBuildOptionDefinitions = function (project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, options) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(options);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "591cb5a4-2d46-4f3a-a697-5cd42b6bd332", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildOptionDefinition, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets report for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} type
    */
    BuildApi.prototype.getBuildReport = function (project, buildId, type) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, report) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(report);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            type: type,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "45bcaa88-67e1-4042-a035-56d3b4a7d44c", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets report for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} type
    */
    BuildApi.prototype.getBuildReportHtmlContent = function (project, buildId, type) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, report) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(report);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            type: type,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "45bcaa88-67e1-4042-a035-56d3b4a7d44c", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("text/html", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    */
    BuildApi.prototype.getResourceUsage = function () {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, ResourceUsage) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(ResourceUsage);
            }
        };
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "3813d06c-9e36-4ea1-aac3-61a485d60e3d", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets revisions of a definition
    *
    * @param {string} project - Project ID or project name
    * @param {number} definitionId
    */
    BuildApi.prototype.getDefinitionRevisions = function (project, definitionId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, revisions) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(revisions);
            }
        };
        var routeValues = {
            project: project,
            definitionId: definitionId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "7c116775-52e5-453e-8c5d-914d9762d8c4", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionRevision, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    */
    BuildApi.prototype.getBuildSettings = function () {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, setting) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(setting);
            }
        };
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "aa8c1c9c-ef8b-474a-b8c4-785c7b191d0d", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Updates the build settings
    *
    * @param {BuildInterfaces.BuildSettings} settings
    */
    BuildApi.prototype.updateBuildSettings = function (settings) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, setting) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(setting);
            }
        };
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "aa8c1c9c-ef8b-474a-b8c4-785c7b191d0d", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.update(url, apiVersion, settings, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Adds a tag to a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} tag
    */
    BuildApi.prototype.addBuildTag = function (project, buildId, tag) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, tags) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(tags);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId,
            tag: tag
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "6e6114b2-8161-44c8-8f6c-c5505782427f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.replace(url, apiVersion, null, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Adds tag to a build
    *
    * @param {string[]} tags
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    */
    BuildApi.prototype.addBuildTags = function (tags, project, buildId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, tags) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(tags);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "6e6114b2-8161-44c8-8f6c-c5505782427f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.create(url, apiVersion, tags, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes a tag from a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} tag
    */
    BuildApi.prototype.deleteBuildTag = function (project, buildId, tag) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, tags) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(tags);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId,
            tag: tag
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "6e6114b2-8161-44c8-8f6c-c5505782427f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the tags for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    */
    BuildApi.prototype.getBuildTags = function (project, buildId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, tags) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(tags);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "6e6114b2-8161-44c8-8f6c-c5505782427f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getTags = function (project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, tags) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(tags);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "d84ac5c6-edc7-43d5-adc9-1b34be5dea09", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes a definition template
    *
    * @param {string} project - Project ID or project name
    * @param {string} templateId
    */
    BuildApi.prototype.deleteTemplate = function (project, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "e884571e-7f92-4d6a-9274-3f5649900835", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Gets definition template filtered by id
    *
    * @param {string} project - Project ID or project name
    * @param {string} templateId
    */
    BuildApi.prototype.getTemplate = function (project, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, template) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(template);
            }
        };
        var routeValues = {
            project: project,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "e884571e-7f92-4d6a-9274-3f5649900835", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionTemplate, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    */
    BuildApi.prototype.getTemplates = function (project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, templates) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(templates);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "e884571e-7f92-4d6a-9274-3f5649900835", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionTemplate, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Saves a definition template
    *
    * @param {BuildInterfaces.BuildDefinitionTemplate} template
    * @param {string} project - Project ID or project name
    * @param {string} templateId
    */
    BuildApi.prototype.saveTemplate = function (template, project, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, template) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(template);
            }
        };
        var routeValues = {
            project: project,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "build", "e884571e-7f92-4d6a-9274-3f5649900835", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionTemplate, responseTypeMetadata: BuildInterfaces.TypeInfo.BuildDefinitionTemplate, responseIsCollection: false };
            _this.restClient.replace(url, apiVersion, template, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets details for a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {string} timelineId
    * @param {number} changeId
    * @param {string} planId
    */
    BuildApi.prototype.getBuildTimeline = function (project, buildId, timelineId, changeId, planId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, Timeline) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(Timeline);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId,
            timelineId: timelineId
        };
        var queryValues = {
            changeId: changeId,
            planId: planId,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "8baac422-4c6e-4de5-8532-db96d92acffa", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: BuildInterfaces.TypeInfo.Timeline, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the work item ids associated with a build
    *
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {number} top - The maximum number of workitems to return
    */
    BuildApi.prototype.getBuildWorkItemsRefs = function (project, buildId, top) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workitems) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workitems);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            '$top': top,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "5a21f5d2-5642-47e4-a0bd-1356e6731bee", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the work item ids associated with build commits
    *
    * @param {string[]} commitIds
    * @param {string} project - Project ID or project name
    * @param {number} buildId
    * @param {number} top - The maximum number of workitems to return, also number of commits to consider if commitids are not sent
    */
    BuildApi.prototype.getBuildWorkItemsRefsFromCommits = function (commitIds, project, buildId, top) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workitems) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workitems);
            }
        };
        var routeValues = {
            project: project,
            buildId: buildId
        };
        var queryValues = {
            '$top': top,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "5a21f5d2-5642-47e4-a0bd-1356e6731bee", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.create(url, apiVersion, commitIds, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets all the work item ids inbetween fromBuildId to toBuildId
    *
    * @param {string} project - Project ID or project name
    * @param {number} fromBuildId
    * @param {number} toBuildId
    * @param {number} top - The maximum number of workitems to return
    */
    BuildApi.prototype.getWorkItemsBetweenBuilds = function (project, fromBuildId, toBuildId, top) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workitems) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workitems);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            fromBuildId: fromBuildId,
            toBuildId: toBuildId,
            '$top': top,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "build", "52ba8915-5518-42e3-a4bb-b0182d159e2d", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    return BuildApi;
}(basem.ClientApiBase));
exports.BuildApi = BuildApi;
