// vso-node-api/WorkItemTrackingApi.js
/*
* ---------------------------------------------------------
* Copyright(C) Microsoft Corporation. All rights reserved.
* ---------------------------------------------------------
*
* ---------------------------------------------------------
* Generated file, DO NOT EDIT
* ---------------------------------------------------------
*/
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
// Licensed under the MIT license.  See LICENSE file in the project root for full license information.
var Q = require('q');
var basem = require('./ClientApiBases');
var WorkItemTrackingInterfaces = require("./interfaces/WorkItemTrackingInterfaces");
var WorkItemTrackingApi = (function (_super) {
    __extends(WorkItemTrackingApi, _super);
    function WorkItemTrackingApi(baseUrl, handlers) {
        _super.call(this, baseUrl, handlers, 'node-WorkItemTracking-api');
    }
    /**
    * Creates an attachment.
    *
    * @param {NodeJS.ReadableStream} contentStream - Content to upload
    * @param {string} fileName
    * @param {string} uploadType
    */
    WorkItemTrackingApi.prototype.createAttachment = function (customHeaders, contentStream, fileName, uploadType) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, attachment) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(attachment);
            }
        };
        var routeValues = {};
        var queryValues = {
            fileName: fileName,
            uploadType: uploadType,
        };
        customHeaders = customHeaders || {};
        customHeaders["Content-Type"] = "application/octet-stream";
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "e07b5fa4-1499-494d-a496-64b860fd64ff", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.uploadStream('POST', url, apiVersion, contentStream, customHeaders, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns an attachment
    *
    * @param {string} id
    * @param {string} fileName
    */
    WorkItemTrackingApi.prototype.getAttachmentContent = function (id, fileName) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, attachment) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(attachment);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            fileName: fileName,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "e07b5fa4-1499-494d-a496-64b860fd64ff", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("application/octet-stream", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns an attachment
    *
    * @param {string} id
    * @param {string} fileName
    */
    WorkItemTrackingApi.prototype.getAttachmentZip = function (id, fileName) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, attachment) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(attachment);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            fileName: fileName,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "e07b5fa4-1499-494d-a496-64b860fd64ff", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            var accept = _this.createAcceptHeader("application/zip", apiVersion);
            _this.httpClient.getStream(url, accept, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {number} depth
    */
    WorkItemTrackingApi.prototype.getRootNodes = function (project, depth) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, classificationNodes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(classificationNodes);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            '$depth': depth,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a70579d1-f53a-48ee-a5be-7be8659023b9", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {WorkItemTrackingInterfaces.WorkItemClassificationNode} postedNode
    * @param {string} project - Project ID or project name
    * @param {WorkItemTrackingInterfaces.TreeStructureGroup} structureGroup
    * @param {string} path
    */
    WorkItemTrackingApi.prototype.createOrUpdateClassificationNode = function (postedNode, project, structureGroup, path) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, classificationNode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(classificationNode);
            }
        };
        var routeValues = {
            project: project,
            structureGroup: structureGroup,
            path: path
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "5a172953-1b41-49d3-840a-33f79c3ce89f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, postedNode, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {WorkItemTrackingInterfaces.TreeStructureGroup} structureGroup
    * @param {string} path
    * @param {number} reclassifyId
    */
    WorkItemTrackingApi.prototype.deleteClassificationNode = function (project, structureGroup, path, reclassifyId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            structureGroup: structureGroup,
            path: path
        };
        var queryValues = {
            '$reclassifyId': reclassifyId,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "5a172953-1b41-49d3-840a-33f79c3ce89f", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {WorkItemTrackingInterfaces.TreeStructureGroup} structureGroup
    * @param {string} path
    * @param {number} depth
    */
    WorkItemTrackingApi.prototype.getClassificationNode = function (project, structureGroup, path, depth) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, classificationNode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(classificationNode);
            }
        };
        var routeValues = {
            project: project,
            structureGroup: structureGroup,
            path: path
        };
        var queryValues = {
            '$depth': depth,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "5a172953-1b41-49d3-840a-33f79c3ce89f", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {WorkItemTrackingInterfaces.WorkItemClassificationNode} postedNode
    * @param {string} project - Project ID or project name
    * @param {WorkItemTrackingInterfaces.TreeStructureGroup} structureGroup
    * @param {string} path
    */
    WorkItemTrackingApi.prototype.updateClassificationNode = function (postedNode, project, structureGroup, path) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, classificationNode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(classificationNode);
            }
        };
        var routeValues = {
            project: project,
            structureGroup: structureGroup,
            path: path
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "5a172953-1b41-49d3-840a-33f79c3ce89f", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemClassificationNode, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, postedNode, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns comment for a work item at the specified revision
    *
    * @param {number} id
    * @param {number} revision
    */
    WorkItemTrackingApi.prototype.getComment = function (id, revision) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, comment) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(comment);
            }
        };
        var routeValues = {
            id: id,
            revision: revision
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "19335ae7-22f7-4308-93d8-261f9384b7cf", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemComment, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns specified number of comments for a work item from the specified revision
    *
    * @param {number} id - Work item id
    * @param {number} fromRevision - Revision from which comments are to be fetched
    * @param {number} top - The number of comments to return
    * @param {WorkItemTrackingInterfaces.CommentSortOrder} order - Ascending or descending by revision id
    */
    WorkItemTrackingApi.prototype.getComments = function (id, fromRevision, top, order) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, comment) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(comment);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            fromRevision: fromRevision,
            '$top': top,
            order: order,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "19335ae7-22f7-4308-93d8-261f9384b7cf", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemComments, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets information on a specific field.
    *
    * @param {string} field - Field name
    */
    WorkItemTrackingApi.prototype.getField = function (field) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, field) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(field);
            }
        };
        var routeValues = {
            field: field
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "b51fd764-e5c2-4b9b-aaf7-3395cf4bdd94", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemField, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns information for all fields.
    *
    * @param {WorkItemTrackingInterfaces.GetFieldsExpand} expand - Use ExtensionFields to include extension fields, otherwise exclude them. Unless the feature flag for this parameter is enabled, extension fields are always included.
    */
    WorkItemTrackingApi.prototype.getFields = function (expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, fields) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(fields);
            }
        };
        var routeValues = {};
        var queryValues = {
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "b51fd764-e5c2-4b9b-aaf7-3395cf4bdd94", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemField, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns history of all revision for a given work item ID
    *
    * @param {number} id
    * @param {number} top
    * @param {number} skip
    */
    WorkItemTrackingApi.prototype.getHistory = function (id, top, skip) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, history) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(history);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f74eba29-47a1-4152-9381-84040aced527", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemHistory, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns the history value of particular revision
    *
    * @param {number} id
    * @param {number} revisionNumber
    */
    WorkItemTrackingApi.prototype.getHistoryById = function (id, revisionNumber) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, history) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(history);
            }
        };
        var routeValues = {
            id: id,
            revisionNumber: revisionNumber
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f74eba29-47a1-4152-9381-84040aced527", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemHistory, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Creates a query, or moves a query.
    *
    * @param {WorkItemTrackingInterfaces.QueryHierarchyItem} postedQuery - The query to create.
    * @param {string} project - Project ID or project name
    * @param {string} query - The parent path for the query to create.
    */
    WorkItemTrackingApi.prototype.createQuery = function (postedQuery, project, query) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, querie) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(querie);
            }
        };
        var routeValues = {
            project: project,
            query: query
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a67d190c-c41f-424b-814d-0e906f659301", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, postedQuery, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {string} query
    */
    WorkItemTrackingApi.prototype.deleteQuery = function (project, query) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            query: query
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a67d190c-c41f-424b-814d-0e906f659301", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Retrieves all queries the user has access to in the current project
    *
    * @param {string} project - Project ID or project name
    * @param {WorkItemTrackingInterfaces.QueryExpand} expand
    * @param {number} depth
    * @param {boolean} includeDeleted
    */
    WorkItemTrackingApi.prototype.getQueries = function (project, expand, depth, includeDeleted) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, queries) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(queries);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            '$expand': expand,
            '$depth': depth,
            '$includeDeleted': includeDeleted,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a67d190c-c41f-424b-814d-0e906f659301", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Retrieves a single query by project and either id or path
    *
    * @param {string} project - Project ID or project name
    * @param {string} query
    * @param {WorkItemTrackingInterfaces.QueryExpand} expand
    * @param {number} depth
    * @param {boolean} includeDeleted
    */
    WorkItemTrackingApi.prototype.getQuery = function (project, query, expand, depth, includeDeleted) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, querie) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(querie);
            }
        };
        var routeValues = {
            project: project,
            query: query
        };
        var queryValues = {
            '$expand': expand,
            '$depth': depth,
            '$includeDeleted': includeDeleted,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a67d190c-c41f-424b-814d-0e906f659301", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {WorkItemTrackingInterfaces.QueryHierarchyItem} queryUpdate
    * @param {string} project - Project ID or project name
    * @param {string} query
    * @param {boolean} undeleteDescendants
    */
    WorkItemTrackingApi.prototype.updateQuery = function (queryUpdate, project, query, undeleteDescendants) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, querie) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(querie);
            }
        };
        var routeValues = {
            project: project,
            query: query
        };
        var queryValues = {
            '$undeleteDescendants': undeleteDescendants,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a67d190c-c41f-424b-814d-0e906f659301", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.QueryHierarchyItem, responseIsCollection: false };
            _this.restClient.update(url, apiVersion, queryUpdate, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {number} id
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.destroyWorkItem = function (id, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {
            project: project,
            id: id
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "b70d8d39-926c-465e-b927-b1bf0e5ca0e0", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * @param {number} id
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.getDeletedWorkItem = function (id, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, recyclebin) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(recyclebin);
            }
        };
        var routeValues = {
            project: project,
            id: id
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "b70d8d39-926c-465e-b927-b1bf0e5ca0e0", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    * @param {number[]} ids
    */
    WorkItemTrackingApi.prototype.getDeletedWorkItems = function (project, ids) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, recyclebin) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(recyclebin);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            ids: ids && ids.join(","),
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "b70d8d39-926c-465e-b927-b1bf0e5ca0e0", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {WorkItemTrackingInterfaces.WorkItemDeleteUpdate} payload
    * @param {number} id
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.restoreWorkItem = function (payload, id, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, recyclebin) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(recyclebin);
            }
        };
        var routeValues = {
            project: project,
            id: id
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "b70d8d39-926c-465e-b927-b1bf0e5ca0e0", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.update(url, apiVersion, payload, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a fully hydrated work item for the requested revision
    *
    * @param {number} id
    * @param {number} revisionNumber
    * @param {WorkItemTrackingInterfaces.WorkItemExpand} expand
    */
    WorkItemTrackingApi.prototype.getRevision = function (id, revisionNumber, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, revision) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(revision);
            }
        };
        var routeValues = {
            id: id,
            revisionNumber: revisionNumber
        };
        var queryValues = {
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a00c85a5-80fa-4565-99c3-bcd2181434bb", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns the list of fully hydrated work item revisions, paged.
    *
    * @param {number} id
    * @param {number} top
    * @param {number} skip
    * @param {WorkItemTrackingInterfaces.WorkItemExpand} expand
    */
    WorkItemTrackingApi.prototype.getRevisions = function (id, top, skip, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, revisions) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(revisions);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            '$top': top,
            '$skip': skip,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a00c85a5-80fa-4565-99c3-bcd2181434bb", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Validates the fields values.
    *
    * @param {WorkItemTrackingInterfaces.FieldsToEvaluate} ruleEngineInput
    */
    WorkItemTrackingApi.prototype.evaluateRulesOnField = function (ruleEngineInput) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "1a3a1536-dca6-4509-b9c3-dd9bb2981506", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.create(url, apiVersion, ruleEngineInput, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Creates a template
    *
    * @param {WorkItemTrackingInterfaces.WorkItemTemplate} template - Template contents
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    */
    WorkItemTrackingApi.prototype.createTemplate = function (template, teamContext) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, template) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(template);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "6a90345f-a676-4969-afce-8e163e1d5642", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.create(url, apiVersion, template, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets template
    *
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {string} workitemtypename - Optional, When specified returns templates for given Work item type.
    */
    WorkItemTrackingApi.prototype.getTemplates = function (teamContext, workitemtypename) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, templates) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(templates);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team
        };
        var queryValues = {
            workitemtypename: workitemtypename,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "6a90345f-a676-4969-afce-8e163e1d5642", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Deletes the template with given id
    *
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {string} templateId - Template id
    */
    WorkItemTrackingApi.prototype.deleteTemplate = function (teamContext, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(null);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "fb10264a-8836-48a0-8033-1b0ccd2748d5", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode);
        });
        return deferred.promise;
    };
    /**
    * Gets the template with specified id
    *
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {string} templateId - Template Id
    */
    WorkItemTrackingApi.prototype.getTemplate = function (teamContext, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, template) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(template);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "fb10264a-8836-48a0-8033-1b0ccd2748d5", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Replace template contents
    *
    * @param {WorkItemTrackingInterfaces.WorkItemTemplate} templateContent - Template contents to replace with
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {string} templateId - Template id
    */
    WorkItemTrackingApi.prototype.replaceTemplate = function (templateContent, teamContext, templateId) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, template) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(template);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team,
            templateId: templateId
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "fb10264a-8836-48a0-8033-1b0ccd2748d5", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.replace(url, apiVersion, templateContent, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a single update for a work item
    *
    * @param {number} id
    * @param {number} updateNumber
    */
    WorkItemTrackingApi.prototype.getUpdate = function (id, updateNumber) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, update) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(update);
            }
        };
        var routeValues = {
            id: id,
            updateNumber: updateNumber
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "6570bf97-d02c-4a91-8d93-3abe9895b1a9", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemUpdate, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a the deltas between work item revisions
    *
    * @param {number} id
    * @param {number} top
    * @param {number} skip
    */
    WorkItemTrackingApi.prototype.getUpdates = function (id, top, skip) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, updates) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(updates);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            '$top': top,
            '$skip': skip,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "6570bf97-d02c-4a91-8d93-3abe9895b1a9", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemUpdate, responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the results of the query.
    *
    * @param {WorkItemTrackingInterfaces.Wiql} wiql - The query containing the wiql.
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {boolean} timePrecision
    * @param {number} top
    */
    WorkItemTrackingApi.prototype.queryByWiql = function (wiql, teamContext, timePrecision, top) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, wiql) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(wiql);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team
        };
        var queryValues = {
            timePrecision: timePrecision,
            '$top': top,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "1a9c53f7-f243-4447-b110-35ef023636e4", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemQueryResult, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, wiql, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the results of the query by id.
    *
    * @param {string} id - The query id.
    * @param {TfsCoreInterfaces.TeamContext} teamContext - The team context for the operation
    * @param {boolean} timePrecision
    */
    WorkItemTrackingApi.prototype.queryById = function (id, teamContext, timePrecision) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, wiql) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(wiql);
            }
        };
        var project = teamContext.projectId || teamContext.project;
        var team = teamContext.teamId || teamContext.team;
        var routeValues = {
            project: project,
            team: team,
            id: id
        };
        var queryValues = {
            timePrecision: timePrecision,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "a02355f5-5f8a-4671-8e32-369d23aac83d", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemQueryResult, responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Get a batch of work item links
    *
    * @param {string} project - Project ID or project name
    * @param {string[]} types - A list of types to filter the results to specific work item types. Omit this parameter to get work item links of all work item types.
    * @param {string} continuationToken - Specifies the continuationToken to start the batch from. Omit this parameter to get the first batch of links.
    * @param {Date} startDateTime - Date/time to use as a starting point for link changes. Only link changes that occurred after that date/time will be returned. Cannot be used in conjunction with 'watermark' parameter.
    */
    WorkItemTrackingApi.prototype.getReportingLinks = function (project, types, continuationToken, startDateTime) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemLink) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemLink);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            types: types && types.join(","),
            continuationToken: continuationToken,
            startDateTime: startDateTime,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "b5b5b6d0-0308-40a1-b3f4-b9bb3c66878f", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Gets the work item relation types.
    *
    * @param {string} relation
    */
    WorkItemTrackingApi.prototype.getRelationType = function (relation) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemRelationType) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemRelationType);
            }
        };
        var routeValues = {
            relation: relation
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f5d33bc9-5b49-4a3c-a9bd-f3cd46dd2165", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    */
    WorkItemTrackingApi.prototype.getRelationTypes = function () {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemRelationTypes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemRelationTypes);
            }
        };
        var routeValues = {};
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f5d33bc9-5b49-4a3c-a9bd-f3cd46dd2165", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Get a batch of work item revisions with the option of including deleted items
    *
    * @param {string} project - Project ID or project name
    * @param {string[]} fields - A list of fields to return in work item revisions. Omit this parameter to get all reportable fields.
    * @param {string[]} types - A list of types to filter the results to specific work item types. Omit this parameter to get work item revisions of all work item types.
    * @param {string} continuationToken - Specifies the watermark to start the batch from. Omit this parameter to get the first batch of revisions.
    * @param {Date} startDateTime - Date/time to use as a starting point for revisions, all revisions will occur after this date/time. Cannot be used in conjunction with 'watermark' parameter.
    * @param {boolean} includeIdentityRef - Return an identity reference instead of a string value for identity fields.
    * @param {boolean} includeDeleted - Specify if the deleted item should be returned.
    * @param {boolean} includeTagRef - Specify if the tag objects should be returned for System.Tags field.
    * @param {boolean} includeLatestOnly - Return only the latest revisions of work items, skipping all historical revisions
    * @param {WorkItemTrackingInterfaces.ReportingRevisionsExpand} expand
    */
    WorkItemTrackingApi.prototype.readReportingRevisionsGet = function (project, fields, types, continuationToken, startDateTime, includeIdentityRef, includeDeleted, includeTagRef, includeLatestOnly, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemRevision) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemRevision);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            fields: fields && fields.join(","),
            types: types && types.join(","),
            continuationToken: continuationToken,
            startDateTime: startDateTime,
            includeIdentityRef: includeIdentityRef,
            includeDeleted: includeDeleted,
            includeTagRef: includeTagRef,
            includeLatestOnly: includeLatestOnly,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f828fe59-dd87-495d-a17c-7a8d6211ca6c", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Get a batch of work item revisions
    *
    * @param {WorkItemTrackingInterfaces.ReportingWorkItemRevisionsFilter} filter - An object that contains request settings: field filter, type filter, identity format
    * @param {string} project - Project ID or project name
    * @param {string} continuationToken - Specifies the watermark to start the batch from. Omit this parameter to get the first batch of revisions.
    * @param {Date} startDateTime - Date/time to use as a starting point for revisions, all revisions will occur after this date/time. Cannot be used in conjunction with 'watermark' parameter.
    * @param {WorkItemTrackingInterfaces.ReportingRevisionsExpand} expand
    */
    WorkItemTrackingApi.prototype.readReportingRevisionsPost = function (filter, project, continuationToken, startDateTime, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemRevision) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemRevision);
            }
        };
        var routeValues = {
            project: project
        };
        var queryValues = {
            continuationToken: continuationToken,
            startDateTime: startDateTime,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "f828fe59-dd87-495d-a17c-7a8d6211ca6c", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.create(url, apiVersion, filter, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {number} id
    * @param {boolean} destroy
    */
    WorkItemTrackingApi.prototype.deleteWorkItem = function (id, destroy) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItem) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItem);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            destroy: destroy,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "72c7ddf8-2cdc-4f60-90cd-ab71c14a399b", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.del(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a single work item
    *
    * @param {number} id
    * @param {string[]} fields
    * @param {Date} asOf
    * @param {WorkItemTrackingInterfaces.WorkItemExpand} expand
    */
    WorkItemTrackingApi.prototype.getWorkItem = function (id, fields, asOf, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItem) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItem);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            fields: fields && fields.join(","),
            asOf: asOf,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "72c7ddf8-2cdc-4f60-90cd-ab71c14a399b", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a list of work items
    *
    * @param {number[]} ids
    * @param {string[]} fields
    * @param {Date} asOf
    * @param {WorkItemTrackingInterfaces.WorkItemExpand} expand
    */
    WorkItemTrackingApi.prototype.getWorkItems = function (ids, fields, asOf, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItems) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItems);
            }
        };
        var routeValues = {};
        var queryValues = {
            ids: ids && ids.join(","),
            fields: fields && fields.join(","),
            asOf: asOf,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "72c7ddf8-2cdc-4f60-90cd-ab71c14a399b", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {VSSInterfaces.JsonPatchDocument} document
    * @param {number} id
    * @param {boolean} validateOnly
    * @param {boolean} bypassRules
    */
    WorkItemTrackingApi.prototype.updateWorkItem = function (customHeaders, document, id, validateOnly, bypassRules) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItem) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItem);
            }
        };
        var routeValues = {
            id: id
        };
        var queryValues = {
            validateOnly: validateOnly,
            bypassRules: bypassRules,
        };
        customHeaders = customHeaders || {};
        customHeaders["Content-Type"] = "application/json-patch+json";
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "72c7ddf8-2cdc-4f60-90cd-ab71c14a399b", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.update(url, apiVersion, document, customHeaders, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {VSSInterfaces.JsonPatchDocument} document
    * @param {string} project - Project ID or project name
    * @param {string} type
    * @param {boolean} validateOnly
    * @param {boolean} bypassRules
    */
    WorkItemTrackingApi.prototype.createWorkItem = function (customHeaders, document, project, type, validateOnly, bypassRules) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItem) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItem);
            }
        };
        var routeValues = {
            project: project,
            type: type
        };
        var queryValues = {
            validateOnly: validateOnly,
            bypassRules: bypassRules,
        };
        customHeaders = customHeaders || {};
        customHeaders["Content-Type"] = "application/json-patch+json";
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "62d3d110-0047-428c-ad3c-4fe872c91c74", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.create(url, apiVersion, document, customHeaders, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a single work item from a template
    *
    * @param {string} project - Project ID or project name
    * @param {string} type
    * @param {string} fields
    * @param {Date} asOf
    * @param {WorkItemTrackingInterfaces.WorkItemExpand} expand
    */
    WorkItemTrackingApi.prototype.getWorkItemTemplate = function (project, type, fields, asOf, expand) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItem) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItem);
            }
        };
        var routeValues = {
            project: project,
            type: type
        };
        var queryValues = {
            fields: fields,
            asOf: asOf,
            '$expand': expand,
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "62d3d110-0047-428c-ad3c-4fe872c91c74", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.getWorkItemTypeCategories = function (project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypeCategories) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypeCategories);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "9b9f5734-36c8-415e-ba67-f83b45c31408", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a the deltas between work item revisions
    *
    * @param {string} project - Project ID or project name
    * @param {string} category
    */
    WorkItemTrackingApi.prototype.getWorkItemTypeCategory = function (project, category) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypeCategorie) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypeCategorie);
            }
        };
        var routeValues = {
            project: project,
            category: category
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "9b9f5734-36c8-415e-ba67-f83b45c31408", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns a the deltas between work item revisions
    *
    * @param {string} project - Project ID or project name
    * @param {string} type
    */
    WorkItemTrackingApi.prototype.getWorkItemType = function (project, type) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemType) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemType);
            }
        };
        var routeValues = {
            project: project,
            type: type
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "7c8d7a76-4a09-43e8-b5df-bd792f4ac6aa", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.getWorkItemTypes = function (project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypes) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypes);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.2", "wit", "7c8d7a76-4a09-43e8-b5df-bd792f4ac6aa", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: true };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Returns the dependent fields for the corresponding workitem type and fieldname
    *
    * @param {string} project - Project ID or project name
    * @param {string} type
    * @param {string} field
    */
    WorkItemTrackingApi.prototype.getDependentFields = function (project, type, field) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypesField) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypesField);
            }
        };
        var routeValues = {
            project: project,
            type: type,
            field: field
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "bd293ce5-3d25-4192-8e67-e8092e879efb", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Export work item type
    *
    * @param {string} project - Project ID or project name
    * @param {string} type
    * @param {boolean} exportGlobalLists
    */
    WorkItemTrackingApi.prototype.exportWorkItemTypeDefinition = function (project, type, exportGlobalLists) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypeTemplate) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypeTemplate);
            }
        };
        var routeValues = {
            project: project,
            type: type
        };
        var queryValues = {
            exportGlobalLists: exportGlobalLists,
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "8637ac8b-5eb6-4f90-b3f7-4f2ff576a459", routeValues, queryValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { responseIsCollection: false };
            _this.restClient.get(url, apiVersion, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    /**
    * Add/updates a work item type
    *
    * @param {WorkItemTrackingInterfaces.WorkItemTypeTemplateUpdateModel} updateModel
    * @param {string} project - Project ID or project name
    */
    WorkItemTrackingApi.prototype.updateWorkItemTypeDefinition = function (updateModel, project) {
        var _this = this;
        var deferred = Q.defer();
        var onResult = function (err, statusCode, workItemTypeTemplate) {
            if (err) {
                err.statusCode = statusCode;
                deferred.reject(err);
            }
            else {
                deferred.resolve(workItemTypeTemplate);
            }
        };
        var routeValues = {
            project: project
        };
        this.vsoClient.getVersioningData("3.0-preview.1", "wit", "8637ac8b-5eb6-4f90-b3f7-4f2ff576a459", routeValues)
            .then(function (versioningData) {
            var url = versioningData.requestUrl;
            var apiVersion = versioningData.apiVersion;
            var serializationData = { requestTypeMetadata: WorkItemTrackingInterfaces.TypeInfo.WorkItemTypeTemplateUpdateModel, responseIsCollection: false };
            _this.restClient.create(url, apiVersion, updateModel, null, serializationData, onResult);
        })
            .fail(function (error) {
            onResult(error, error.statusCode, null);
        });
        return deferred.promise;
    };
    return WorkItemTrackingApi;
}(basem.ClientApiBase));
exports.WorkItemTrackingApi = WorkItemTrackingApi;
